{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1ea90887-1e8f-4ba3-b010-039a30c2be9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jislam/.local/lib/python3.10/site-packages/gosip/__init__.py:5: FutureWarning: In the future `np.ulong` will be defined as the corresponding NumPy scalar.\n",
      "  if not hasattr(np, 'ulong'):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "\t\t\t<script type=\"text/javascript\">\n",
       "\t\t\t<!--\n",
       "\t\t\t\t\n",
       "\t\t\t{\n",
       "\t\t\t\tvar element = document.getElementById('NetworKit_script');\n",
       "\t\t\t\tif (element) {\n",
       "\t\t\t\t\telement.parentNode.removeChild(element);\n",
       "\t\t\t\t}\n",
       "\t\t\t\telement = document.createElement('script');\n",
       "\t\t\t\telement.type = 'text/javascript';\n",
       "\t\t\t\telement.innerHTML = 'function NetworKit_pageEmbed(id) { var i, j; var elements; elements = document.getElementById(id).getElementsByClassName(\"Plot\"); for (i=0; i<elements.length; i++) { elements[i].id = id + \"_Plot_\" + i; var data = elements[i].getAttribute(\"data-image\").split(\"|\"); elements[i].removeAttribute(\"data-image\"); var content = \"<div class=\\\\\"Image\\\\\" id=\\\\\"\" + elements[i].id + \"_Image\\\\\" />\"; elements[i].innerHTML = content; elements[i].setAttribute(\"data-image-index\", 0); elements[i].setAttribute(\"data-image-length\", data.length); for (j=0; j<data.length; j++) { elements[i].setAttribute(\"data-image-\" + j, data[j]); } NetworKit_plotUpdate(elements[i]); elements[i].onclick = function (e) { NetworKit_overlayShow((e.target) ? e.target : e.srcElement); } } elements = document.getElementById(id).getElementsByClassName(\"HeatCell\"); for (i=0; i<elements.length; i++) { var data = parseFloat(elements[i].getAttribute(\"data-heat\")); var color = \"#00FF00\"; if (data <= 1 && data > 0) { color = \"hsla(0, 100%, 75%, \" + (data) + \")\"; } else if (data <= 0 && data >= -1) { color = \"hsla(240, 100%, 75%, \" + (-data) + \")\"; } elements[i].style.backgroundColor = color; } elements = document.getElementById(id).getElementsByClassName(\"Details\"); for (i=0; i<elements.length; i++) { elements[i].setAttribute(\"data-title\", \"-\"); NetworKit_toggleDetails(elements[i]); elements[i].onclick = function (e) { NetworKit_toggleDetails((e.target) ? e.target : e.srcElement); } } elements = document.getElementById(id).getElementsByClassName(\"MathValue\"); for (i=elements.length-1; i>=0; i--) { value = elements[i].innerHTML.trim(); if (value === \"nan\") { elements[i].parentNode.innerHTML = \"\" } } elements = document.getElementById(id).getElementsByClassName(\"SubCategory\"); for (i=elements.length-1; i>=0; i--) { value = elements[i].innerHTML.trim(); if (value === \"\") { elements[i].parentNode.removeChild(elements[i]) } } elements = document.getElementById(id).getElementsByClassName(\"Category\"); for (i=elements.length-1; i>=0; i--) { value = elements[i].innerHTML.trim(); if (value === \"\") { elements[i].parentNode.removeChild(elements[i]) } } var isFirefox = false; try { isFirefox = typeof InstallTrigger !== \"undefined\"; } catch (e) {} if (!isFirefox) { alert(\"Currently the function\\'s output is only fully supported by Firefox.\"); } } function NetworKit_plotUpdate(source) { var index = source.getAttribute(\"data-image-index\"); var data = source.getAttribute(\"data-image-\" + index); var image = document.getElementById(source.id + \"_Image\"); image.style.backgroundImage = \"url(\" + data + \")\"; } function NetworKit_showElement(id, show) { var element = document.getElementById(id); element.style.display = (show) ? \"block\" : \"none\"; } function NetworKit_overlayShow(source) { NetworKit_overlayUpdate(source); NetworKit_showElement(\"NetworKit_Overlay\", true); } function NetworKit_overlayUpdate(source) { document.getElementById(\"NetworKit_Overlay_Title\").innerHTML = source.title; var index = source.getAttribute(\"data-image-index\"); var data = source.getAttribute(\"data-image-\" + index); var image = document.getElementById(\"NetworKit_Overlay_Image\"); image.setAttribute(\"data-id\", source.id); image.style.backgroundImage = \"url(\" + data + \")\"; var link = document.getElementById(\"NetworKit_Overlay_Toolbar_Bottom_Save\"); link.href = data; link.download = source.title + \".svg\"; } function NetworKit_overlayImageShift(delta) { var image = document.getElementById(\"NetworKit_Overlay_Image\"); var source = document.getElementById(image.getAttribute(\"data-id\")); var index = parseInt(source.getAttribute(\"data-image-index\")); var length = parseInt(source.getAttribute(\"data-image-length\")); var index = (index+delta) % length; if (index < 0) { index = length + index; } source.setAttribute(\"data-image-index\", index); NetworKit_overlayUpdate(source); } function NetworKit_toggleDetails(source) { var childs = source.children; var show = false; if (source.getAttribute(\"data-title\") == \"-\") { source.setAttribute(\"data-title\", \"+\"); show = false; } else { source.setAttribute(\"data-title\", \"-\"); show = true; } for (i=0; i<childs.length; i++) { if (show) { childs[i].style.display = \"block\"; } else { childs[i].style.display = \"none\"; } } }';\n",
       "\t\t\t\telement.setAttribute('id', 'NetworKit_script');\n",
       "\t\t\t\tdocument.head.appendChild(element);\n",
       "\t\t\t}\n",
       "\t\t\n",
       "\t\t\t\t\n",
       "\t\t\t{\n",
       "\t\t\t\tvar element = document.getElementById('NetworKit_style');\n",
       "\t\t\t\tif (element) {\n",
       "\t\t\t\t\telement.parentNode.removeChild(element);\n",
       "\t\t\t\t}\n",
       "\t\t\t\telement = document.createElement('style');\n",
       "\t\t\t\telement.type = 'text/css';\n",
       "\t\t\t\telement.innerHTML = '.NetworKit_Page { font-family: Arial, Helvetica, sans-serif; font-size: 14px; } .NetworKit_Page .Value:before { font-family: Arial, Helvetica, sans-serif; font-size: 1.05em; content: attr(data-title) \":\"; margin-left: -2.5em; padding-right: 0.5em; } .NetworKit_Page .Details .Value:before { display: block; } .NetworKit_Page .Value { font-family: monospace; white-space: pre; padding-left: 2.5em; white-space: -moz-pre-wrap !important; white-space: -pre-wrap; white-space: -o-pre-wrap; white-space: pre-wrap; word-wrap: break-word; tab-size: 4; -moz-tab-size: 4; } .NetworKit_Page .Category { clear: both; padding-left: 1em; margin-bottom: 1.5em; } .NetworKit_Page .Category:before { content: attr(data-title); font-size: 1.75em; display: block; margin-left: -0.8em; margin-bottom: 0.5em; } .NetworKit_Page .SubCategory { margin-bottom: 1.5em; padding-left: 1em; } .NetworKit_Page .SubCategory:before { font-size: 1.6em; display: block; margin-left: -0.8em; margin-bottom: 0.5em; } .NetworKit_Page .SubCategory[data-title]:before { content: attr(data-title); } .NetworKit_Page .Block { display: block; } .NetworKit_Page .Block:after { content: \".\"; visibility: hidden; display: block; height: 0; clear: both; } .NetworKit_Page .Block .Thumbnail_Overview, .NetworKit_Page .Block .Thumbnail_ScatterPlot { width: 260px; float: left; } .NetworKit_Page .Block .Thumbnail_Overview img, .NetworKit_Page .Block .Thumbnail_ScatterPlot img { width: 260px; } .NetworKit_Page .Block .Thumbnail_Overview:before, .NetworKit_Page .Block .Thumbnail_ScatterPlot:before { display: block; text-align: center; font-weight: bold; } .NetworKit_Page .Block .Thumbnail_Overview:before { content: attr(data-title); } .NetworKit_Page .HeatCell { font-family: \"Courier New\", Courier, monospace; cursor: pointer; } .NetworKit_Page .HeatCell, .NetworKit_Page .HeatCellName { display: inline; padding: 0.1em; margin-right: 2px; background-color: #FFFFFF } .NetworKit_Page .HeatCellName { margin-left: 0.25em; } .NetworKit_Page .HeatCell:before { content: attr(data-heat); display: inline-block; color: #000000; width: 4em; text-align: center; } .NetworKit_Page .Measure { clear: both; } .NetworKit_Page .Measure .Details { cursor: pointer; } .NetworKit_Page .Measure .Details:before { content: \"[\" attr(data-title) \"]\"; display: block; } .NetworKit_Page .Measure .Details .Value { border-left: 1px dotted black; margin-left: 0.4em; padding-left: 3.5em; pointer-events: none; } .NetworKit_Page .Measure .Details .Spacer:before { content: \".\"; opacity: 0.0; pointer-events: none; } .NetworKit_Page .Measure .Plot { width: 440px; height: 440px; cursor: pointer; float: left; margin-left: -0.9em; margin-right: 20px; } .NetworKit_Page .Measure .Plot .Image { background-repeat: no-repeat; background-position: center center; background-size: contain; height: 100%; pointer-events: none; } .NetworKit_Page .Measure .Stat { width: 500px; float: left; } .NetworKit_Page .Measure .Stat .Group { padding-left: 1.25em; margin-bottom: 0.75em; } .NetworKit_Page .Measure .Stat .Group .Title { font-size: 1.1em; display: block; margin-bottom: 0.3em; margin-left: -0.75em; border-right-style: dotted; border-right-width: 1px; border-bottom-style: dotted; border-bottom-width: 1px; background-color: #D0D0D0; padding-left: 0.2em; } .NetworKit_Page .Measure .Stat .Group .List { -webkit-column-count: 3; -moz-column-count: 3; column-count: 3; } .NetworKit_Page .Measure .Stat .Group .List .Entry { position: relative; line-height: 1.75em; } .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:before { position: absolute; left: 0; top: -40px; background-color: #808080; color: #ffffff; height: 30px; line-height: 30px; border-radius: 5px; padding: 0 15px; content: attr(data-tooltip); white-space: nowrap; display: none; } .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:after { position: absolute; left: 15px; top: -10px; border-top: 7px solid #808080; border-left: 7px solid transparent; border-right: 7px solid transparent; content: \"\"; display: none; } .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:hover:after, .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:hover:before { display: block; } .NetworKit_Page .Measure .Stat .Group .List .Entry .MathValue { font-family: \"Courier New\", Courier, monospace; } .NetworKit_Page .Measure:after { content: \".\"; visibility: hidden; display: block; height: 0; clear: both; } .NetworKit_Page .PartitionPie { clear: both; } .NetworKit_Page .PartitionPie img { width: 600px; } #NetworKit_Overlay { left: 0px; top: 0px; display: none; position: absolute; width: 100%; height: 100%; background-color: rgba(0,0,0,0.6); z-index: 1000; } #NetworKit_Overlay_Title { position: absolute; color: white; transform: rotate(-90deg); width: 32em; height: 32em; padding-right: 0.5em; padding-top: 0.5em; text-align: right; font-size: 40px; } #NetworKit_Overlay .button { background: white; cursor: pointer; } #NetworKit_Overlay .button:before { size: 13px; display: inline-block; text-align: center; margin-top: 0.5em; margin-bottom: 0.5em; width: 1.5em; height: 1.5em; } #NetworKit_Overlay .icon-close:before { content: \"X\"; } #NetworKit_Overlay .icon-previous:before { content: \"P\"; } #NetworKit_Overlay .icon-next:before { content: \"N\"; } #NetworKit_Overlay .icon-save:before { content: \"S\"; } #NetworKit_Overlay_Toolbar_Top, #NetworKit_Overlay_Toolbar_Bottom { position: absolute; width: 40px; right: 13px; text-align: right; z-index: 1100; } #NetworKit_Overlay_Toolbar_Top { top: 0.5em; } #NetworKit_Overlay_Toolbar_Bottom { Bottom: 0.5em; } #NetworKit_Overlay_ImageContainer { position: absolute; top: 5%; left: 5%; height: 90%; width: 90%; background-repeat: no-repeat; background-position: center center; background-size: contain; } #NetworKit_Overlay_Image { height: 100%; width: 100%; background-repeat: no-repeat; background-position: center center; background-size: contain; }';\n",
       "\t\t\t\telement.setAttribute('id', 'NetworKit_style');\n",
       "\t\t\t\tdocument.head.appendChild(element);\n",
       "\t\t\t}\n",
       "\t\t\n",
       "\t\t\t\t\n",
       "\t\t\t{\n",
       "\t\t\t\tvar element = document.getElementById('NetworKit_Overlay');\n",
       "\t\t\t\tif (element) {\n",
       "\t\t\t\t\telement.parentNode.removeChild(element);\n",
       "\t\t\t\t}\n",
       "\t\t\t\telement = document.createElement('div');\n",
       "\t\t\t\telement.innerHTML = '<div id=\"NetworKit_Overlay_Toolbar_Top\"><div class=\"button icon-close\" id=\"NetworKit_Overlay_Close\" /></div><div id=\"NetworKit_Overlay_Title\" /> <div id=\"NetworKit_Overlay_ImageContainer\"> <div id=\"NetworKit_Overlay_Image\" /> </div> <div id=\"NetworKit_Overlay_Toolbar_Bottom\"> <div class=\"button icon-previous\" onclick=\"NetworKit_overlayImageShift(-1)\" /> <div class=\"button icon-next\" onclick=\"NetworKit_overlayImageShift(1)\" /> <a id=\"NetworKit_Overlay_Toolbar_Bottom_Save\"><div class=\"button icon-save\" /></a> </div>';\n",
       "\t\t\t\telement.setAttribute('id', 'NetworKit_Overlay');\n",
       "\t\t\t\tdocument.body.appendChild(element);\n",
       "\t\t\t\tdocument.getElementById('NetworKit_Overlay_Close').onclick = function (e) {\n",
       "\t\t\t\t\tdocument.getElementById('NetworKit_Overlay').style.display = 'none';\n",
       "\t\t\t\t}\n",
       "\t\t\t}\n",
       "\t\t\n",
       "\t\t\t-->\n",
       "\t\t\t</script>\n",
       "\t\t"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pickle\n",
    "from distutils.dir_util import copy_tree\n",
    "with open('/home/jislam/Documents/gits/reproduce_thesis/chapter_02/extra_scripts/src.py') as f:\n",
    "    exec(f.read())\n",
    "torch.manual_seed(1)\n",
    "\n",
    "# Check if GPU is available\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "sns.set(style=\"whitegrid\")\n",
    "#from lifelines import CoxPHFitter\n",
    "#from captum.attr import DeepLiftShap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a24462a3-0562-4ba7-ac7b-3359ecfd8a0a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_247180/698340387.py:40: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs[\"cell_disease\"]=adata.obs[\"cell_type\"].astype(str) + '_' + adata.obs[\"disease_state\"].astype(str)\n",
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "min_optuna=10\n",
    "hyperparam_epochs=10 #max per trial(note that we add burn in inside the loop\n",
    "hyperparam_trials=100 #number of trials for each model.\n",
    "patience=5\n",
    "burn_in=10\n",
    "num_epochs=1000\n",
    "k=10\n",
    "bsize=2048\n",
    "validation_ratio=0.1 #10% of dataset is kept for validation during training. we do not use a testing set since we're only interested in what it learned not generalizability.\n",
    "\n",
    "\n",
    "main_path=\"/home/jislam/Documents/gits/reproduce_thesis/chapter_02/analyses/\"\n",
    "dataLocation=\"/home/jislam/Documents/gits/scRNAseq_datasets/hpap_batch_corrected.final3k.h5ad\"\n",
    "filter_by=\"cell_type\"\n",
    "categories_to_keep = 8 #number of cell types I want to keep in the model sorted by decreasing cell count, can use all but only so much weighted losses can do...\n",
    "category_labels=[\"cell_disease\"] # See adata.obs.columns for what categories you can use here. Specify which columns to one-hot-encode and use to inform the models. cell_disease is generated for the interaction.\n",
    "\n",
    "group_a=[\"beta_cell_T2D\"]\n",
    "group_b=[\"beta_cell_Control\"]\n",
    "shared_filter =None\n",
    "\n",
    "single_cell_data=\"hpap\"\n",
    "\n",
    "\n",
    "\n",
    "outdir=single_cell_data.replace(\" \", \"_\")\n",
    "if not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)\n",
    "\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "adata=sc.read_h5ad(dataLocation)\n",
    "adata.obs['cell_type']=adata.obs['cell_type'].str.replace(\" \", \"_\", regex=False)\n",
    "adata=adata[adata.obs[\"disease_state\"]!=\"AAB\"]\n",
    "adata=adata[adata.obs[\"disease_state\"]!=\"T1D\"]\n",
    "adata=adata[adata.obs[\"cell_type\"]!=\"unknown\"]\n",
    "adata= adata[adata.obs['cell_type'].isin(['alpha_cell', 'delta_cell', 'beta_cell', 'PP_cell','acinar_cell'])]#, 'precursor_alpha_beta_cell'\n",
    "#adata.obs.loc[adata.obs['cell_type']=='precursor_alpha_beta_cell','cell_type']=\"precursor_cell\"\n",
    "\n",
    "adata.obs[\"cell_disease\"]=adata.obs[\"cell_type\"].astype(str) + '_' + adata.obs[\"disease_state\"].astype(str)\n",
    "\n",
    "one_hot_labels, num_categories= gosip.one_hot_encode_combinations(adata.obs, category_labels)\n",
    "adata.one_hot_labels= one_hot_labels\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "adata.obs['cell_disease'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80916d85-7170-4637-9fb3-f0bc702f246a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fde7799-9814-4b62-af39-923bf77bcacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "322fe6e1-fd7c-466d-88c2-f7116d32847e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# get dense matrix\n",
    "X = adata.X.toarray() if hasattr(adata.X, \"toarray\") else adata.X\n",
    "\n",
    "# gene‐wise moments\n",
    "μ = X.mean(axis=0)\n",
    "σ2 = X.var(axis=0)\n",
    "# skip genes with no overdispersion\n",
    "#mask = σ2 > μ\n",
    "#μ, σ2 = μ[mask], σ2[mask]\n",
    "\n",
    "# NB parameters\n",
    "r = μ**2 / (σ2 - μ)\n",
    "p = r / (r + μ)\n",
    "\n",
    "# expected vs observed zero‐fractions\n",
    "exp_zero = p**r\n",
    "obs_zero = (X == 0).mean(axis=0)\n",
    "\n",
    "# which genes exceed expectation?\n",
    "inflated = obs_zero > exp_zero\n",
    "print(f\"{inflated.mean()*100:.1f}% of genes have more zeros than expected\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d064e7c-4364-4440-ab7e-d474ca970f48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1674f7eb-a2bc-496d-85cd-66faa4679070",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(adata[adata.obs['cell_disease']==\"delta_cell_Control\"].X[:,adata.var.index==\"SST\"]).hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14d64e2-5159-4dc9-aa4c-b21834e142de",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "gene_pick='INS'\n",
    "\n",
    "print(pd.DataFrame(adata.X[adata.obs['cell_disease']=='beta_cell_Control',adata.var.index==gene_pick]).mean())\n",
    "print(pd.DataFrame(adata.X[adata.obs['cell_disease']=='beta_cell_T2D',adata.var.index==gene_pick]).mean())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b616507-0525-4daf-8359-1649eaac53af",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X[adata.obs['cell_disease']=='beta_cell_control',adata.var.index==gene_pick]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a22315-32bf-452a-bcff-14af987ae50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.tl.rank_genes_groups(adata, groupby='cell_disease', groups=group_a, reference=group_b[0], method='t-test', use_raw=False,rankby_abs=True,key_added=\"cure_T2D\")\n",
    "#sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False,save=outdir+\"basic_test.png\")\n",
    "sc.tl.rank_genes_groups(adata, groupby='cell_disease', groups=group_b, reference=group_a[0], method='t-test', use_raw=False,rankby_abs=True,key_added=\"cause_T2D\")\n",
    "#sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False,save=outdir+\"basic_test.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c8c521c-4d3d-4326-8da9-1683f187c381",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43e9de21-817c-447e-9fa2-255172b7d676",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=[\"donor_id\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca6bc6b-e382-4b3a-9233-a2adeb1b4f95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the column data\n",
    "column_data = adata.obs[filter_by]\n",
    "\n",
    "# Count the occurrences of each unique value\n",
    "value_counts = column_data.value_counts()\n",
    "value_counts.index=[ value_counts.index[v]+\"=\"+str(value_counts[v])  for v in range(value_counts.shape[0])]\n",
    "# Create the pie chart\n",
    "plt.figure(figsize=(8, 8))\n",
    "plt.pie(value_counts, labels=value_counts.index, autopct='%.0f%%', startangle=-30)\n",
    "plt.title(filter_by)\n",
    "plt.show()\n",
    "\n",
    "# Extract the column data\n",
    "column_data = adata.obs['cell_disease']\n",
    "\n",
    "# Count the occurrences of each unique value\n",
    "value_counts = column_data.value_counts()\n",
    "value_counts.index=[ value_counts.index[v]+\"=\"+str(value_counts[v])  for v in range(value_counts.shape[0])]\n",
    "# Create the pie chart\n",
    "plt.figure(figsize=(8, 8))\n",
    "plt.pie(value_counts, labels=value_counts.index, autopct='%.0f%%', startangle=-30)\n",
    "plt.title('Cell x disease interaction')\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1abfbfa4-9113-49ca-8c54-cd88ad973648",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bba09012-8d33-43fa-a158-fff992ddc0cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "(adata.X==0).sum(0).sum()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06590a52-d995-4b14-a086-363634915a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#simulation_results={}\n",
    "min_layer_size_star=int(adata.shape[1]*2/8)\n",
    "max_layer_size_star=int(adata.shape[1]*2/4)\n",
    "min_layer_size_oracle=int(adata.shape[1]*2/8)\n",
    "max_layer_size_oracle=int(adata.shape[1]*2/4)\n",
    "\n",
    "min_layer_size_prop=int(adata.shape[1]*2/8)\n",
    "max_layer_size_prop=int(adata.shape[1]*2/4)\n",
    "\n",
    "\n",
    "\n",
    "from copy import deepcopy\n",
    "\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "###########################################################################################\n",
    "###########################################################################################\n",
    "#### Stargan\n",
    "###########################################################################################\n",
    "###########################################################################################         \n",
    "study = optuna.create_study(storage=\"sqlite:///\"+outdir+\"/\"+single_cell_data.replace(\" \",\"_\")+\"_SIMULATION_stargan.db\",\n",
    "                            direction='minimize',\n",
    "                            sampler=optuna.samplers.TPESampler(),\n",
    "                            pruner=optuna.pruners.HyperbandPruner(min_resource=min_optuna,\n",
    "                                                                  max_resource=hyperparam_epochs,\n",
    "                                                                  reduction_factor=5),\n",
    "                            study_name=single_cell_data.replace(\" \", \"_\")+\"_stargan_SIMULATION_\"+\"_main.db\",\n",
    "                            load_if_exists=True)\n",
    "\n",
    "\n",
    "start_time=time.time()\n",
    "if len(study.get_trials())<hyperparam_trials:\n",
    "    study.optimize(lambda trial: opt_objective_stargan(trial, adata,num_categories,hyperparam_epochs,validation_ratio=validation_ratio,min_layer_size=min_layer_size_star,max_layer_size=max_layer_size_star, bsize=bsize), n_trials=hyperparam_trials-len(study.get_trials()))\n",
    "\n",
    "# Get the best hyperparameters\n",
    "best_params = study.best_params\n",
    "\n",
    "stargan_hyperparam = {\n",
    "    'batch_size': best_params['batch_size'],\n",
    "    'learning_rate': best_params['learning_rate'],\n",
    "    'layer_nodes_generator': [\n",
    "        best_params[\"gen_layer0\"],\n",
    "        best_params[\"gen_layer1\"],\n",
    "        best_params[\"gen_layer2\"]\n",
    "    ],\n",
    "    'layer_nodes_discriminator': [\n",
    "        best_params[\"disc_layer0\"],\n",
    "        best_params[\"disc_layer1\"],\n",
    "        best_params[\"disc_layer2\"]\n",
    "    ],\n",
    "    'critics': 3,\n",
    "#\"lambda_adv\":best_params[\"lambda_adv\"],\n",
    "#\"lambda_cls\":best_params[\"lambda_cls\"],\n",
    "#\"lambda_rec\":best_params[\"lambda_rec\"],\n",
    "#\"lambda_iden\":best_params[\"lambda_iden\"],\n",
    "    \"dropout_rate\":best_params[\"dropout_rate\"]\n",
    "}\n",
    "'''\n",
    "stargan_hyperparam = {\n",
    "    'batch_size': 512,\n",
    "    'learning_rate': 0.001,\n",
    "    'layer_nodes_generator': [\n",
    "        250,\n",
    "        100,\n",
    "        250\n",
    "    ],\n",
    "    'layer_nodes_discriminator': [\n",
    "        250,\n",
    "        125,\n",
    "        250\n",
    "    ],\n",
    "    'critics': 5,\n",
    "\"lambda_adv\":1,\n",
    "\"lambda_cls\":1,\n",
    "\"lambda_rec\":10,\n",
    "\"lambda_iden\":1,\n",
    "\"lambda_sink\":1,\n",
    "    \"dropout_rate\":0.1\n",
    "}\n",
    "\n",
    "\n",
    "'''\n",
    "###########################################################################################\n",
    "###########################################################################################\n",
    "#### Oracle\n",
    "###########################################################################################\n",
    "###########################################################################################\n",
    "\n",
    "\n",
    "study = optuna.create_study(storage=\"sqlite:///\"+outdir+\"/\"+single_cell_data.replace(\" \",\"_\")+\"_SIMULATION_oracle.db\",\n",
    "                            direction='minimize',\n",
    "                            sampler=optuna.samplers.TPESampler(),\n",
    "                            pruner=optuna.pruners.HyperbandPruner(min_resource=min_optuna,\n",
    "                                                                  max_resource=hyperparam_epochs,\n",
    "                                                                  reduction_factor=5),\n",
    "                            study_name=single_cell_data.replace(\" \", \"_\")+\"_oracle_SIMULATION_\"+\"main.db\",\n",
    "                            load_if_exists=True)\n",
    "\n",
    "start_time=time.time()\n",
    "if len(study.get_trials())<hyperparam_trials:\n",
    "    study.optimize(lambda trial: opt_objective_oracle(trial, adata,num_categories=num_categories,hyperparam_epochs=hyperparam_epochs,validation_ratio=validation_ratio,min_layer_size=min_layer_size_oracle,max_layer_size=max_layer_size_oracle, bsize=bsize), n_trials=hyperparam_trials-len(study.get_trials()))\n",
    "print(time.time()-start_time)\n",
    "# Get the best hyperparameters\n",
    "best_params = study.best_params\n",
    "print(\"Best hyperparameters:\", best_params)\n",
    "oracle_hyperparam = {'batch_size': best_params['batch_size'],\n",
    "    'learning_rate': best_params['learning_rate'],\n",
    "    'layer_nodes': [\n",
    "        best_params[\"disc_layer0\"],\n",
    "        best_params[\"disc_layer1\"],\n",
    "        best_params[\"disc_layer2\"]\n",
    "    ],\n",
    "    'dropout_rate_d': best_params['disc_dropout_rate']\n",
    "}\n",
    "'''\n",
    "\n",
    "oracle_hyperparam = {'batch_size': 512,\n",
    "    'learning_rate': 0.001,\n",
    "    'layer_nodes': [\n",
    "        200,\n",
    "        150,\n",
    "        100\n",
    "    ],\n",
    "    'dropout_rate_d': 0.1\n",
    "}\n",
    "'''       \n",
    "###########################################################################################\n",
    "###########################################################################################\n",
    "#### Propagator\n",
    "###########################################################################################\n",
    "###########################################################################################\n",
    "\n",
    "study = optuna.create_study(storage=\"sqlite:///\"+outdir+\"/\"+single_cell_data.replace(\" \",\"_\")+\"_SIMULATION_propagator.db\",\n",
    "                            direction='minimize',\n",
    "                            sampler=optuna.samplers.TPESampler(),\n",
    "                            pruner=optuna.pruners.HyperbandPruner(min_resource=min_optuna,\n",
    "                                                                  max_resource=hyperparam_epochs,\n",
    "                                                                  reduction_factor=5),\n",
    "                            study_name=single_cell_data.replace(\" \", \"_\")+\"_propagator_SIMULATION_\"+\"main.db\",\n",
    "                            load_if_exists=True)\n",
    "\n",
    "start_time=time.time()\n",
    "if len(study.get_trials())<hyperparam_trials:\n",
    "    study.optimize(lambda trial: opt_objective_propagator(trial, adata,num_categories=num_categories,hyperparam_epochs=hyperparam_epochs,validation_ratio=validation_ratio,min_layer_size=min_layer_size_prop,max_layer_size=max_layer_size_prop, bsize=bsize), n_trials=hyperparam_trials-len(study.get_trials()))\n",
    "print(time.time()-start_time)\n",
    "# Get the best hyperparameters\n",
    "best_params = study.best_params\n",
    "print(\"Best hyperparameters:\", best_params)\n",
    "\n",
    "propagator_hyperparam = {\n",
    "    'batch_size': best_params['batch_size'],\n",
    "    'learning_rate': best_params['learning_rate'],\n",
    "    'layer_nodes_generator': [\n",
    "        best_params[\"gen_layer0\"],\n",
    "        best_params[\"gen_layer1\"],\n",
    "        best_params[\"gen_layer2\"]\n",
    "    ],\n",
    "    'dropout_rate_g': best_params['gen_dropout_rate'],\n",
    "    'latent_dim':best_params['latent'],\n",
    "    'beta':float(best_params['beta'])\n",
    "    \n",
    "}\n",
    "\n",
    "'''\n",
    "propagator_hyperparam = {\n",
    "    'batch_size': 512,\n",
    "    'learning_rate': 0.001,\n",
    "    'layer_nodes_generator': [\n",
    "        250,\n",
    "        100,\n",
    "        250\n",
    "    ],\n",
    "    'dropout_rate_g': 0.1,\n",
    "    'latent_dim':50,\n",
    "    'beta':10.0\n",
    "}\n",
    "'''\n",
    "shared_filter = None\n",
    "filter_criteria_start = [\"start\"]\n",
    "filter_criteria_goal  = [\"goal\"]\n",
    "\n",
    "\n",
    "\n",
    "loss_weights=[1,1,10,10,10]\n",
    "stargan = gosip.StarGAN(input_dim=adata.shape[1],\n",
    "                  num_domains=[adata.one_hot_labels.shape[1]],\n",
    "                  device = device,\n",
    "                  learning_rate = stargan_hyperparam['learning_rate'],\n",
    "                  layer_g = stargan_hyperparam['layer_nodes_generator'],\n",
    "                  layer_d = stargan_hyperparam['layer_nodes_discriminator'],\n",
    "                  critics = stargan_hyperparam['critics'],\n",
    "                  lambda_adv=loss_weights[0],\n",
    "                  lambda_cls=loss_weights[1],\n",
    "                  lambda_rec=loss_weights[2],\n",
    "                  #lambda_iden=loss_weights[3],\n",
    "                  dropout_rate=stargan_hyperparam[\"dropout_rate\"])\n",
    "\n",
    "train_adata,val_adata=  split_and_maxabs_scale_adata(adata,validation_ratio)\n",
    "train_dataloader = gosip.prepare_data(train_adata,num_categories,batch_size=stargan_hyperparam[\"batch_size\"])\n",
    "val_dataloader = gosip.prepare_data(val_adata,num_categories,batch_size=stargan_hyperparam[\"batch_size\"])\n",
    "stargan.train( dataloader=train_dataloader,val_loader=val_dataloader,num_epochs=num_epochs,patience=patience,burn_in=burn_in,verbose=True)\n",
    "stargan.G.eval()\n",
    "stargan.D.eval()\n",
    "torch.cuda.empty_cache()\n",
    "\n",
    "oracle = gosip.Oracle(input_dim=adata.shape[1],\n",
    "         num_domains=num_categories,\n",
    "         device = device,\n",
    "         learning_rate = oracle_hyperparam['learning_rate'],\n",
    "         layer_d = oracle_hyperparam['layer_nodes'],\n",
    "         drpt_d = oracle_hyperparam['dropout_rate_d'])\n",
    "train_dataloader = gosip.prepare_data(train_adata,num_categories,batch_size=oracle_hyperparam[\"batch_size\"])\n",
    "val_dataloader = gosip.prepare_data(val_adata,num_categories,batch_size=oracle_hyperparam[\"batch_size\"])\n",
    "oracle.train(dataloader=train_dataloader, val_loader=val_dataloader,num_epochs=num_epochs,patience=patience,burn_in=burn_in, verbose=True,optuna_run=False)\n",
    "oracle.D.eval()\n",
    "torch.cuda.empty_cache()\n",
    "\n",
    "propagator = gosip.Propagator(input_dim=train_adata.shape[1],\n",
    "                 num_domains=num_categories,\n",
    "                 device = device,\n",
    "                 learning_rate = propagator_hyperparam['learning_rate'],\n",
    "                 layer_g = propagator_hyperparam['layer_nodes_generator'],\n",
    "                 drpt_g = propagator_hyperparam['dropout_rate_g'],\n",
    "                 latent_dim=propagator_hyperparam['latent_dim'])\n",
    "loss_fn = gosip.BtcvaeLoss(n_data=train_adata.shape[0], alpha=1.0, beta=propagator_hyperparam['beta'], gamma=1.0)\n",
    "train_dataloader = gosip.prepare_data(train_adata,num_categories,batch_size=propagator_hyperparam[\"batch_size\"])\n",
    "val_dataloader = gosip.prepare_data(val_adata,num_categories,batch_size=propagator_hyperparam[\"batch_size\"])\n",
    "propagator.train(dataloader=train_dataloader, val_loader=val_dataloader,\n",
    "                 num_epochs=num_epochs,patience=patience,burn_in=burn_in, verbose=True,optuna_run=False,\n",
    "                 loss_fn=loss_fn)\n",
    "\n",
    "\n",
    "\n",
    "propagator.G.eval()\n",
    "torch.cuda.empty_cache()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "clear_output(wait=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba7d7212-c45e-43f6-b392-6f7923bbef8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02e54ced-f017-4d87-9ffc-4b5d140e0490",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('hpap/stargan_hyperparam_selected.json', 'w') as f:\n",
    "    json.dump(stargan_hyperparam, f)\n",
    "\n",
    "with open('hpap/propagator_hyperparam_selected.json', 'w') as f:\n",
    "    json.dump(propagator_hyperparam, f)\n",
    "\n",
    "with open('hpap/oracle_hyperparam_selected.json', 'w') as f:\n",
    "    json.dump(oracle_hyperparam, f)\n",
    "propagator_hyperparameter_ranges = optuna.create_study(storage=\"sqlite:///\"+outdir+\"/\"+single_cell_data.replace(\" \",\"_\")+\"_SIMULATION_propagator.db\",\n",
    "                            direction='minimize',\n",
    "                            sampler=optuna.samplers.TPESampler(),\n",
    "                            pruner=optuna.pruners.HyperbandPruner(min_resource=min_optuna,\n",
    "                                                                  max_resource=hyperparam_epochs,\n",
    "                                                                  reduction_factor=5),\n",
    "                            study_name=single_cell_data.replace(\" \", \"_\")+\"_propagator_SIMULATION_\"+\"main.db\",\n",
    "                            load_if_exists=True).trials[0].distributions\n",
    "\n",
    "oracle_hyperparameter_ranges = optuna.create_study(storage=\"sqlite:///\"+outdir+\"/\"+single_cell_data.replace(\" \",\"_\")+\"_SIMULATION_oracle.db\",\n",
    "                            direction='minimize',\n",
    "                            sampler=optuna.samplers.TPESampler(),\n",
    "                            pruner=optuna.pruners.HyperbandPruner(min_resource=min_optuna,\n",
    "                                                                  max_resource=hyperparam_epochs,\n",
    "                                                                  reduction_factor=5),\n",
    "                            study_name=single_cell_data.replace(\" \", \"_\")+\"_oracle_SIMULATION_\"+\"main.db\",\n",
    "                            load_if_exists=True).trials[0].distributions\n",
    "stargan_hyperparameter_ranges = optuna.create_study(storage=\"sqlite:///\"+outdir+\"/\"+single_cell_data.replace(\" \",\"_\")+\"_SIMULATION_stargan.db\",\n",
    "                            direction='minimize',\n",
    "                            sampler=optuna.samplers.TPESampler(),\n",
    "                            pruner=optuna.pruners.HyperbandPruner(min_resource=min_optuna,\n",
    "                                                                  max_resource=hyperparam_epochs,\n",
    "                                                                  reduction_factor=5),\n",
    "                            study_name=single_cell_data.replace(\" \", \"_\")+\"_stargan_SIMULATION_\"+\"_main.db\",\n",
    "                            load_if_exists=True).trials[0].distributions\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd05aa6b-b8a4-4839-a95e-525acaacf242",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def convert_distributions_to_string(input_dict):\n",
    "    # Create a new dictionary to store the updated values\n",
    "    output_dict = {}\n",
    "    \n",
    "    for key, value in input_dict.items():\n",
    "        output_dict[key] = str(value)  # for other types, store the original value\n",
    "\n",
    "    return output_dict\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "import json\n",
    "with open('hpap/stargan_hyperparam_ranges.json', 'w') as f:\n",
    "    json.dump(convert_distributions_to_string(stargan_hyperparameter_ranges), f)\n",
    "\n",
    "with open('hpap/propagator_hyperparam_ranges.json', 'w') as f:\n",
    "    json.dump(convert_distributions_to_string(propagator_hyperparameter_ranges), f)\n",
    "\n",
    "with open('hpap/oracle_hyperparam_ranges.json', 'w') as f:\n",
    "    json.dump(convert_distributions_to_string(oracle_hyperparameter_ranges), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52089a9e-8204-4c64-909f-e427e28f559e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['disease_state']==\"Control\"\n",
    "adata.obs['cell disease']=adata.obs['cell_disease'].str.replace(\"_\",\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc1e1b22-2b2c-411a-a7f5-30fc0380ebe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(adata.obs['donor_id'].value_counts().index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d7ff087-cfa1-4bb5-9ab1-c4efb0cf0103",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Set figure size\n",
    "fig = plt.figure(figsize=(13, 13))\n",
    "gs = fig.add_gridspec(2, 2, height_ratios=[1, 0.8], width_ratios=[1, 1], figure=fig)\n",
    "\n",
    "# Function to add labels on bars\n",
    "def add_bar_labels(ax, values):\n",
    "    for i, v in enumerate(values):\n",
    "        ax.text(v + max(values) * 0.02, i, str(v), va='center', fontsize=10)  # Shift label slightly to the right\n",
    "\n",
    "# Function to add subplot labels (A, B, C)\n",
    "def add_subplot_label(ax, label,x_pos=0.02,y_pos=0.99):\n",
    "    ax.text(x_pos,y_pos, label, transform=ax.transAxes, fontsize=14, fontweight='bold', va='top', ha='left')\n",
    "\n",
    "# --- PLOT 1: UMAP (Spanning the full first row) ---\n",
    "ax1 = fig.add_subplot(gs[0, :])  # Full row for UMAP\n",
    "sc.pl.umap(adata, color=[\"cell disease\"], ax=ax1, show=False)\n",
    "ax1.set_title(\"UMAP of Cell x Disease\")\n",
    "add_subplot_label(ax1, \"A\",x_pos=0.02,y_pos=0.98)  # Add label \"A\"\n",
    "\n",
    "# --- PLOT 2: Bar Chart for Control State ---\n",
    "ax2 = fig.add_subplot(gs[1, 0])\n",
    "column_data = adata.obs.loc[adata.obs['disease_state'] == \"Control\", 'cell_type']\n",
    "value_counts = column_data.value_counts(ascending=True)\n",
    "\n",
    "sns.barplot(x=value_counts.values, y=value_counts.index.str.replace(\"_\", \" \"), ax=ax2, palette=\"Blues_r\")\n",
    "ax2.set_title(\"Control State: Cell Type Distribution\")\n",
    "ax2.set_xlabel(\"Cell Count\")\n",
    "ax2.set_ylabel(\"Cell Type\")\n",
    "add_bar_labels(ax2, value_counts.values)\n",
    "add_subplot_label(ax2, \"B\")  # Add label \"B\"\n",
    "\n",
    "# --- PLOT 3: Bar Chart for T2D State ---\n",
    "ax3 = fig.add_subplot(gs[1, 1])\n",
    "column_data = adata.obs.loc[adata.obs['disease_state'] == \"T2D\", 'cell_type']\n",
    "value_counts = column_data.value_counts(ascending=True)\n",
    "\n",
    "# Ensure the last element is removed safely\n",
    "if value_counts.shape[0] > 1:\n",
    "    value_counts = value_counts\n",
    "\n",
    "sns.barplot(x=value_counts.values, y=value_counts.index.str.replace(\"_\", \" \"), ax=ax3, palette=\"Reds_r\")\n",
    "ax3.set_title(\"T2D State: Cell Type Distribution\")\n",
    "ax3.set_xlabel(\"Cell Count\")\n",
    "ax3.set_ylabel(\"Cell Type\")\n",
    "add_bar_labels(ax3, value_counts.values)\n",
    "add_subplot_label(ax3, \"C\")  # Add label \"C\"\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "fig.savefig('hpap/hpap-developmental-merge.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d08795b-538e-4a90-a0ee-d6556868ea7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('hpap/stargan_hyperparam.json', 'w') as f:\n",
    "    json.dump(stargan_hyperparam, f)\n",
    "\n",
    "with open('hpap/propagator_hyperparam.json', 'w') as f:\n",
    "    json.dump(propagator_hyperparam, f)\n",
    "\n",
    "with open('hpap/oracle_hyperparam.json', 'w') as f:\n",
    "    json.dump(oracle_hyperparam, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7302cb03-c96a-4671-a125-aecb052646e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.tl.rank_genes_groups(adata, groupby='cell_disease', groups=['alpha_cell_T2D'], reference='alpha_cell_Control', method='t-test', use_raw=False,rankby_abs=True,key_added=\"cause_alpha_T2D\")\n",
    "sc.pl.rank_genes_groups(adata, n_genes=50,key=\"cause_alpha_T2D\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "124055cd-b822-49d5-81ab-1be0c61356d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.tl.rank_genes_groups(adata, groupby='cell_disease', groups=group_a, reference=group_b[0], method='t-test', use_raw=False,rankby_abs=True,key_added=\"cure_T2D\")\n",
    "#sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False,save=outdir+\"basic_test.png\")\n",
    "sc.tl.rank_genes_groups(adata, groupby='cell_disease', groups=group_b, reference=group_a[0], method='t-test', use_raw=False,rankby_abs=True,key_added=\"cause_T2D\")\n",
    "#sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False,save=outdir+\"basic_test.png\")\n",
    "cure_dge = extract_rank_genes_groups(adata, key='cure_T2D', n_genes=None).drop(columns=['logfoldchange','pval','pval_adj','group'])\n",
    "cause_dge = extract_rank_genes_groups(adata, key='cause_T2D', n_genes=None).drop(columns=['logfoldchange','pval','pval_adj','group'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd731b2e-fcf0-4d24-b294-48b381ba6115",
   "metadata": {},
   "outputs": [],
   "source": [
    "processed_adata = anndata.concat(\n",
    "    [train_adata, val_adata],\n",
    "    join='inner',            # only keep genes present in both\n",
    "    label='split',           # name of the new obs column\n",
    "    keys=['train', 'val'],   # values in combined_adata.obs['split']\n",
    "    fill_value=0\n",
    ")\n",
    "\n",
    "# 2. Stitch the one_hot_labels together\n",
    "combined_labels = pd.concat(\n",
    "    [train_adata.one_hot_labels, val_adata.one_hot_labels],\n",
    "    axis=0\n",
    ")\n",
    "# 3. Re‐order them to match the new obs order, then attach back\n",
    "combined_labels = combined_labels.reindex(processed_adata.obs_names)\n",
    "processed_adata.one_hot_labels = combined_labels.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cca77cb-3286-45ae-954b-70951e34c341",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e26b8aa0-36cb-4c7e-aafd-293712d148f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "start_time=time.time()\n",
    "shared_filter = None\n",
    "filter_criteria_start = [\"alpha_cell_Control\"] #[\"cell_disease=beta cell_Tumor\"] #\n",
    "filter_criteria_goal =[\"alpha_cell_T2D\"] # [\"cell_disease=beta cell_Normal\"]# \n",
    "perturbation_metrics,accuracy,result_path=gosip.full_report(adata=processed_adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                propagator=propagator,\n",
    "                                shared_filter=shared_filter,\n",
    "                                filter_criteria_start=filter_criteria_start,\n",
    "                                filter_criteria_goal=filter_criteria_goal,\n",
    "                                device=device,num_categories=num_categories,category_labels=category_labels,\n",
    "                                main_path=main_path,outdir=outdir,top_n=20,percentile=0,\n",
    "                                umap=False,oracle_performance=True,alpha=0.5,k=k)\n",
    "\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()\n",
    "\n",
    "perturbation_metrics.to_csv(result_path+\"summary_statistics.csv\")\n",
    "perturbation_metrics_start_to_goal=perturbation_metrics.copy()\n",
    "perturbation_metrics,accuracy,result_path=gosip.full_report(adata=processed_adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                          propagator=propagator,\n",
    "                                          shared_filter=shared_filter,\n",
    "                                 filter_criteria_start=filter_criteria_goal,\n",
    "                                 filter_criteria_goal=filter_criteria_start,\n",
    "                                device=device,num_categories=num_categories,category_labels=category_labels,\n",
    "                                 main_path=main_path,outdir=outdir,top_n=20,percentile=0,\n",
    "                                 umap=False,oracle_performance=True,alpha=0.5,k=k)\n",
    "perturbation_metrics.to_csv(result_path+\"summary_statistics.csv\")\n",
    "perturbation_metrics_goal_to_start=perturbation_metrics.copy()\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "268d028a-c0da-4531-b8f1-745986d6dc11",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "shared_filter = None\n",
    "filter_criteria_start = [\"delta_cell_Control\"] #[\"cell_disease=beta cell_Tumor\"] #\n",
    "filter_criteria_goal =[\"delta_cell_T2D\"] # [\"cell_disease=beta cell_Normal\"]# \n",
    "perturbation_metrics,accuracy,result_path=gosip.full_report(adata=processed_adata,\n",
    "                                stargan=stargan,\n",
    "                                oracle=oracle,\n",
    "                                propagator=propagator,\n",
    "                                shared_filter=shared_filter,\n",
    "                                filter_criteria_start=filter_criteria_start,\n",
    "                                filter_criteria_goal=filter_criteria_goal,\n",
    "                                device=device,num_categories=num_categories,category_labels=category_labels,\n",
    "                                main_path=main_path,outdir=outdir,top_n=20,percentile=0,\n",
    "                                umap=False,oracle_performance=True,alpha=0.5,k=k)\n",
    "\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()\n",
    "\n",
    "perturbation_metrics.to_csv(result_path+\"summary_statistics.csv\")\n",
    "perturbation_metrics_start_to_goal=perturbation_metrics.copy()\n",
    "perturbation_metrics,accuracy,result_path=gosip.full_report(adata=processed_adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                          propagator=propagator,\n",
    "                                          shared_filter=shared_filter,\n",
    "                                 filter_criteria_start=filter_criteria_goal,\n",
    "                                 filter_criteria_goal=filter_criteria_start,\n",
    "                                device=device,num_categories=num_categories,category_labels=category_labels,\n",
    "                                 main_path=main_path,outdir=outdir,top_n=20,percentile=0,\n",
    "                                 umap=False,oracle_performance=True,alpha=0.5,k=k)\n",
    "perturbation_metrics.to_csv(result_path+\"summary_statistics.csv\")\n",
    "perturbation_metrics_goal_to_start=perturbation_metrics.copy()\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c44f6351-316a-439e-b087-f741265e445e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "shared_filter = None\n",
    "filter_criteria_start = [\"beta_cell_Control\"] #[\"cell_disease=beta cell_Tumor\"] #\n",
    "filter_criteria_goal =[\"beta_cell_T2D\"] # [\"cell_disease=beta cell_Normal\"]# \n",
    "perturbation_metrics,accuracy,result_path=gosip.full_report(adata=processed_adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                propagator=propagator,\n",
    "                                shared_filter=shared_filter,\n",
    "                                filter_criteria_start=filter_criteria_start,\n",
    "                                filter_criteria_goal=filter_criteria_goal,\n",
    "                                device=device,num_categories=num_categories,category_labels=category_labels,\n",
    "                                main_path=main_path,outdir=outdir,top_n=20,percentile=0,\n",
    "                                umap=False,oracle_performance=True,alpha=0.5,k=k)\n",
    "\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()\n",
    "\n",
    "perturbation_metrics.to_csv(result_path+\"summary_statistics.csv\")\n",
    "perturbation_metrics_start_to_goal=perturbation_metrics.copy()\n",
    "perturbation_metrics,accuracy,result_path=gosip.full_report(adata=processed_adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                          propagator=propagator,\n",
    "                                          shared_filter=shared_filter,\n",
    "                                 filter_criteria_start=filter_criteria_goal,\n",
    "                                 filter_criteria_goal=filter_criteria_start,\n",
    "                                device=device,num_categories=num_categories,category_labels=category_labels,\n",
    "                                 main_path=main_path,outdir=outdir,top_n=20,percentile=0,\n",
    "                                 umap=False,oracle_performance=True,alpha=0.5,k=k)\n",
    "perturbation_metrics.to_csv(result_path+\"summary_statistics.csv\")\n",
    "perturbation_metrics_goal_to_start=perturbation_metrics.copy()\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()\n",
    "end_time=time.time()\n",
    "print(f'{end_time-start_time:.3f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28bfddcb-e81a-4715-bb13-1507845d7b91",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, groupby='cell_disease', groups=group_a, reference=group_b[0], method='t-test', use_raw=False,rankby_abs=True,key_added=\"cure_T2D\")\n",
    "#sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False,save=outdir+\"basic_test.png\")\n",
    "sc.tl.rank_genes_groups(adata, groupby='cell_disease', groups=group_b, reference=group_a[0], method='t-test', use_raw=False,rankby_abs=True,key_added=\"cause_T2D\")\n",
    "#sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False,save=outdir+\"basic_test.png\")\n",
    "cure_dge = extract_rank_genes_groups(adata, key='cure_T2D', n_genes=None).drop(columns=['logfoldchange','pval','pval_adj','group'])\n",
    "cause_dge = extract_rank_genes_groups(adata, key='cause_T2D', n_genes=None).drop(columns=['logfoldchange','pval','pval_adj','group'])\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6030c2b-deb5-4e8f-8c1b-bda820026ef6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc1cf946-f659-4619-a38c-e75c1ef64db3",
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbation_metrics#.sort_values('oracle_score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "949c6a00-03c3-42ee-8316-849ad1945c15",
   "metadata": {},
   "outputs": [],
   "source": [
    "mart_main=pd.read_csv('/home/jislam/Documents/gits/rnaseq_bulk/pdac_bulk/mart_export.txt',sep=\"\\t\")\n",
    "mart_main=mart_main.loc[~mart_main['NCBI gene (formerly Entrezgene) ID'].isna(),:]\n",
    "mart_main=mart_main.drop(columns=['EntrezGene transcript name ID','Gene stable ID'],axis=1)\n",
    "mart_main=mart_main[~mart_main.duplicated()]\n",
    "mart_main.columns=['gene_name','ncbi_id']\n",
    "mart_main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2c2b0f5-f165-43be-88e9-2288f111fccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.decomposition import PCA\n",
    "import umap.umap_ as umap\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "def plot_umap(dataframe, key_column, highlight_keys=None, figsize=(10, 8), default_color=\"gray\"):\n",
    "    \"\"\"\n",
    "    Plot UMAP-transformed data, highlighting specified keys with colors.\n",
    "\n",
    "    Parameters:\n",
    "    - dataframe (pd.DataFrame): DataFrame containing UMAP components and the key column.\n",
    "    - key_column (str): Column name used for grouping and coloring points.\n",
    "    - highlight_keys (list): List of keys to highlight in distinct colors. Others will be in a default color.\n",
    "    - figsize (tuple): Figure size for the plot (default: (10, 8)).\n",
    "    - default_color (str): Color for keys not in the highlight list (default: \"gray\").\n",
    "    \n",
    "    Returns:\n",
    "    - None: Displays the scatter plot.\n",
    "    \"\"\"\n",
    "    # Validate highlight_keys\n",
    "    highlight_keys = set(highlight_keys) if highlight_keys else set()\n",
    "\n",
    "    # Separate highlighted and non-highlighted data\n",
    "    highlighted_data = dataframe[dataframe[key_column].isin(highlight_keys)]\n",
    "    non_highlighted_data = dataframe[~dataframe[key_column].isin(highlight_keys)]\n",
    "\n",
    "    # Start plotting\n",
    "    plt.figure(figsize=figsize)\n",
    "\n",
    "    # Plot non-highlighted data\n",
    "    if not non_highlighted_data.empty:\n",
    "        plt.scatter(\n",
    "            non_highlighted_data[\"UMAP1\"],\n",
    "            non_highlighted_data[\"UMAP2\"],\n",
    "            color=default_color,\n",
    "            alpha=0.6,\n",
    "            s=1,\n",
    "            label=\"Other\",\n",
    "        )\n",
    "\n",
    "    # Plot highlighted data\n",
    "    if not highlighted_data.empty:\n",
    "        sns.scatterplot(\n",
    "            data=highlighted_data,\n",
    "            x=\"UMAP1\",\n",
    "            y=\"UMAP2\",\n",
    "            hue=key_column,\n",
    "            palette=\"tab10\",  # Use a color palette for highlighted keys\n",
    "            s=5,\n",
    "            alpha=0.9,\n",
    "        )\n",
    "\n",
    "    # Add labels and legend\n",
    "    plt.title(\"UMAP Projection\", fontsize=14)\n",
    "    plt.xlabel(\"UMAP1\", fontsize=12)\n",
    "    plt.ylabel(\"UMAP2\", fontsize=12)\n",
    "    plt.legend(title=key_column, bbox_to_anchor=(1.05, 1), loc=\"upper left\", fontsize=10)\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "    return fig\n",
    "def pca_umap_transform(dataframe, key_column, n_pca_components=40, n_umap_components=2, n_neighbors=5, min_dist=0.3, random_state=None):\n",
    "    \"\"\"\n",
    "    Perform PCA followed by UMAP on a DataFrame, excluding a specified key column.\n",
    "\n",
    "    Parameters:\n",
    "    - dataframe (pd.DataFrame): Input DataFrame.\n",
    "    - key_column (str): Column name to exclude from PCA and UMAP transformations.\n",
    "    - n_pca_components (int): Number of components for PCA.\n",
    "    - n_umap_components (int): Number of dimensions for UMAP reduction.\n",
    "    - n_neighbors (int): Number of neighbors for UMAP.\n",
    "    - min_dist (float): Minimum distance between points in UMAP.\n",
    "    - random_state (int): Seed for reproducibility.\n",
    "\n",
    "    Returns:\n",
    "    - pd.DataFrame: Transformed DataFrame with UMAP components and the key column.\n",
    "    \"\"\"\n",
    "    # Separate the key column\n",
    "    keys = dataframe[key_column]\n",
    "    numerical_data = dataframe.drop(columns=[key_column])\n",
    "    '''\n",
    "    # Apply PCA\n",
    "    pca = PCA(n_components=n_pca_components)\n",
    "    pca_result = pca.fit_transform(numerical_data)\n",
    "    '''\n",
    "    # Apply UMAP\n",
    "    umap_model = umap.UMAP(\n",
    "        n_neighbors=n_neighbors,\n",
    "        min_dist=min_dist,\n",
    "        n_components=n_umap_components,\n",
    "        random_state=random_state,\n",
    "    )\n",
    "    umap_result = umap_model.fit_transform(numerical_data)\n",
    "\n",
    "    # Create a new DataFrame with UMAP results\n",
    "    result_df = pd.DataFrame(umap_result, columns=[f\"UMAP{i+1}\" for i in range(n_umap_components)])\n",
    "    result_df[key_column] = keys.reset_index(drop=True)\n",
    "\n",
    "    return result_df\n",
    "\n",
    "\n",
    "def generate_example_cells(adata,start_tensor, transition_label_tensor, stargan, propagator,cells_to_generate,top_genes,latent):\n",
    "    if latent:\n",
    "        pg=propagator.G.encoder\n",
    "    else:\n",
    "        pg=propagator.G\n",
    "        \n",
    "    cells_set={}\n",
    "    indices = torch.randperm(start_tensor.size(0))[:cells_to_generate]\n",
    "    \n",
    "    device = stargan.device\n",
    "    #here i'm passing the start data through the propagator as well.\n",
    "    td = start_tensor.to(device).clone()#pg()\n",
    "    p_td=pg(td)\n",
    "    \n",
    "    ##maybe i want the raw data through the propagator as well? td and gtd\n",
    "    ## for now, i think it might be a bad idea because true labels aren't perfect\n",
    "    #or, its a good idea because the gene impact without changes should be monitored\n",
    "    ############\n",
    "\n",
    "    gtd = stargan.G(start_tensor.to(device), transition_label_tensor.to(device))\n",
    "    p_gtd=pg(gtd)\n",
    "    cells_set['start']=p_td.clone().detach().cpu().numpy().copy()\n",
    "    cells_set['goal']=p_gtd.clone().detach().cpu().numpy().copy()\n",
    "    td = start_tensor[indices].to(device).clone()#pg()\n",
    "    p_td=propagator.G(td)\n",
    "    gtd = stargan.G(start_tensor[indices].to(device), transition_label_tensor[indices].to(device))\n",
    "    p_gtd=propagator.G(gtd)\n",
    "    with torch.no_grad():\n",
    "        for col_idx in range(start_tensor.shape[1]):  # range(10):\n",
    "            if adata.var.index[col_idx] in top_genes:\n",
    "                temp = p_td[:, col_idx].clone()\n",
    "                p_td[:, col_idx] = p_gtd[:, col_idx].clone()  # Replace column in X_tensor with corresponding column in Y_tensor\n",
    "                mod_pert = pg(p_td)\n",
    "                #mod_pert[:, col_idx] = p_gtd[:, col_idx].clone()\n",
    "                cells_set[adata.var.index[col_idx]]=mod_pert.detach().cpu().numpy().copy()\n",
    "                p_td[:, col_idx] = temp\n",
    "                \n",
    "                clear_output(wait=True)\n",
    "            torch.cuda.empty_cache()\n",
    "    \n",
    "    return cells_set#filter_average_gene_impact(new_pred_ave, p_value, 0.05), t_stat, p_value\n",
    "\n",
    "# Convert dictionary to a single DataFrame\n",
    "def dictionaries_to_pd_df(data):\n",
    "    df_list = []\n",
    "    for key, array in data.items():\n",
    "        temp_df = pd.DataFrame(array)  # Convert array to DataFrame\n",
    "        temp_df['key'] = key  # Add the key as a new column\n",
    "        df_list.append(temp_df)\n",
    "    \n",
    "    # Concatenate all DataFrames\n",
    "    result_df = pd.concat(df_list, ignore_index=True)\n",
    "    return result_df\n",
    "    \n",
    "def generate_examples(adata,stargan,oracle,propagator,shared_filter,filter_criteria_start,filter_criteria_goal,device=None,*,cells_to_generate=None,top_genes=\"nothing\",latent=False):\n",
    "    \n",
    "    torch.cuda.empty_cache()\n",
    "    stargan.D.eval()\n",
    "    stargan.G.eval()\n",
    "    oracle.D.eval()\n",
    "    propagator.G.eval()\n",
    "\n",
    "    \n",
    "    shared_adata=adata.copy()\n",
    "    shared_adata.one_hot_labels=adata.one_hot_labels.copy()\n",
    "    shared_filter=\"None\"\n",
    "    \n",
    "    with torch.no_grad():\n",
    "        start_adata,start_state_tensor, start_label_tensor = gosip.process_adata_to_tensors(shared_adata,1, filter_criteria_start)\n",
    "        goal_adata,goal_state_tensor, goal_label_tensor = gosip.process_adata_to_tensors(shared_adata,1, filter_criteria_goal)\n",
    "        transition_labels=start_adata.one_hot_labels.copy()\n",
    "        transition_labels[start_adata.one_hot_labels.iloc[:,(~start_adata.one_hot_labels.columns.str.contains(filter_criteria_start[0]) & start_adata.one_hot_labels.columns.str.contains(filter_criteria_goal[0]))].columns] = (start_adata.one_hot_labels.iloc[:,start_adata.one_hot_labels.columns.str.contains(filter_criteria_start[0])])\n",
    "        transition_labels[(start_adata.one_hot_labels.iloc[:,start_adata.one_hot_labels.columns.str.contains(filter_criteria_start[0])].columns)] = 0\n",
    "        transition_label_tensor=torch.tensor(transition_labels.values).float()\n",
    "    \n",
    "    \n",
    "        inverse_transition_labels=goal_adata.one_hot_labels.copy()\n",
    "        inverse_transition_labels[goal_adata.one_hot_labels.iloc[:,(~goal_adata.one_hot_labels.columns.str.contains(filter_criteria_goal[0]) & start_adata.one_hot_labels.columns.str.contains(filter_criteria_start[0]))].columns] = (goal_adata.one_hot_labels.iloc[:,goal_adata.one_hot_labels.columns.str.contains(filter_criteria_goal[0])])\n",
    "        inverse_transition_labels[(goal_adata.one_hot_labels.iloc[:,goal_adata.one_hot_labels.columns.str.contains(filter_criteria_goal[0])].columns)] = 0\n",
    "        inverse_transition_label_tensor=torch.tensor(inverse_transition_labels.values).float()\n",
    "    \n",
    "        if cells_to_generate is None:\n",
    "            cells_to_generate=start_state_tensor.shape[0]\n",
    "        cells=generate_example_cells(adata=adata,start_tensor=start_state_tensor,#,stargan.G(start_state_tensor.to(device), torch.tensor(inverse_transition_labels.values).to(device)\n",
    "                                              transition_label_tensor=transition_label_tensor,\n",
    "                                              stargan=stargan,\n",
    "                                             propagator=propagator,cells_to_generate=cells_to_generate,top_genes=top_genes,latent=latent)\n",
    "        return dictionaries_to_pd_df(cells)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68bf46d5-5963-4cc3-a966-f615e194164d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "propagated_cells=generate_examples(adata=adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                 propagator=propagator,\n",
    "                                 shared_filter=shared_filter,\n",
    "                                 filter_criteria_start=filter_criteria_start,\n",
    "                                 filter_criteria_goal=filter_criteria_goal,\n",
    "                                 device=device,cells_to_generate=2000,\n",
    "                                 top_genes=list(perturbation_metrics.sort_values(\"oracle_score\",ascending=False).head(3)['label'])+list(perturbation_metrics.sort_values(\"oracle_score\",ascending=False).tail(3)['label'])+[\"PPY\",\"DKK3\"],\n",
    "                                 latent=False\n",
    "                                 )\n",
    "\n",
    "print(propagated_cells.shape)\n",
    "umap_data=pca_umap_transform(dataframe=propagated_cells, key_column='key', n_pca_components=25, n_umap_components=2, n_neighbors=5, min_dist=1)\n",
    "\n",
    "plot_umap(\n",
    "    dataframe=umap_data,\n",
    "    key_column=\"key\",\n",
    "    highlight_keys=[\"PPY\",\"INS\",\"DKK3\"]+['start','goal'],  # Highlight key1 and key2\n",
    "    figsize=(8, 8),\n",
    "    default_color=\"lightgray\",\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "197442cf-7b9a-439e-9d48-9c330a9b1e8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def plot_density(df, value_col, category_col):\n",
    "    \"\"\"\n",
    "    Plots a density plot for a numerical column in a DataFrame,\n",
    "    with colors assigned based on a categorical column.\n",
    "\n",
    "    Parameters:\n",
    "    df (pd.DataFrame): The input DataFrame.\n",
    "    value_col (str): The name of the numerical column to plot.\n",
    "    category_col (str): The name of the categorical column for coloring.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "    plt.figure(figsize=(8, 5))\n",
    "    sns.kdeplot(data=df, x=value_col, hue=category_col, fill=True, alpha=0.5)\n",
    "    \n",
    "    # Labels and title\n",
    "    plt.xlabel(value_col)\n",
    "    plt.ylabel(\"Density\")\n",
    "    plt.title(f\"Density Plot of {value_col} by {category_col}\")\n",
    "    # Show the plot\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def plot_density_grid(df1, df2, value_cols, category_col, figsize=(16, 12), title1=\"Title 1\", title2=\"Title 2\"):\n",
    "    \"\"\"\n",
    "    Plots a side-by-side grid of density plots for multiple numerical columns in two DataFrames,\n",
    "    with colors assigned based on a categorical column.\n",
    "\n",
    "    Parameters:\n",
    "    df1 (pd.DataFrame): The first input DataFrame.\n",
    "    df2 (pd.DataFrame): The second input DataFrame.\n",
    "    value_cols (list): List of numerical column names to plot.\n",
    "    category_col (str): The categorical column for coloring.\n",
    "    figsize (tuple): Size of the overall figure.\n",
    "    title1 (str): Title for the first dataset plots.\n",
    "    title2 (str): Title for the second dataset plots.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "    rows = len(value_cols)  # One row per gene\n",
    "    fig, axes = plt.subplots(rows, 2, figsize=figsize, constrained_layout=True)\n",
    "\n",
    "    for i, value_col in enumerate(value_cols):\n",
    "        # First column (df1)\n",
    "        sns.kdeplot(\n",
    "            data=df1[df1[category_col].isin([\"start\", \"goal\", value_col])], \n",
    "            x=value_col, hue=category_col, fill=True, alpha=0.5, ax=axes[i, 0]\n",
    "        )\n",
    "        axes[i, 0].set_xlabel(\"\")\n",
    "        axes[i, 0].set_ylabel(value_col+\" density\")\n",
    "        if i == 0:\n",
    "            axes[i, 0].set_title(title1, fontweight=\"bold\")\n",
    "\n",
    "        # Second column (df2)\n",
    "        sns.kdeplot(\n",
    "            data=df2[df2[category_col].isin([\"start\", \"goal\", value_col])], \n",
    "            x=value_col, hue=category_col, fill=True, alpha=0.5, ax=axes[i, 1]\n",
    "        )\n",
    "        axes[i, 1].set_xlabel(\"\")\n",
    "        axes[i, 1].set_ylabel(value_col+\" density\")\n",
    "        if i == 0:\n",
    "            axes[i, 1].set_title(title2, fontweight=\"bold\")\n",
    "    axes[ len(value_cols)-1,0].set_xlabel(\"Value\")\n",
    "    axes[ len(value_cols)-1,1].set_xlabel(\"Value\")\n",
    "    plt.show()\n",
    "    return fig\n",
    "\n",
    "\n",
    "# Gene list\n",
    "gene_viz_set = [\"PPY\", \"PPP1R1A\", \"INS\", \"DKK3\"]\n",
    "\n",
    "\n",
    "# Generate first set of propagated cells (Control to T2D)\n",
    "propagated_cells_1 = generate_examples(\n",
    "    adata=adata,\n",
    "    stargan=stargan,\n",
    "    oracle=oracle,\n",
    "    propagator=propagator,\n",
    "    shared_filter=shared_filter,\n",
    "    filter_criteria_start=filter_criteria_start,\n",
    "    filter_criteria_goal=filter_criteria_goal,\n",
    "    device=device,\n",
    "    cells_to_generate=None,\n",
    "    top_genes=gene_viz_set,\n",
    "    latent=False\n",
    ")\n",
    "propagated_cells_1.columns = list(adata.var.index.copy()) + [\"perturbation\"]\n",
    "\n",
    "# Generate second set of propagated cells (T2D to Control)\n",
    "propagated_cells_2 = generate_examples(\n",
    "    adata=adata,\n",
    "    stargan=stargan,\n",
    "    oracle=oracle,\n",
    "    propagator=propagator,\n",
    "    shared_filter=shared_filter,\n",
    "    filter_criteria_start=filter_criteria_goal,\n",
    "    filter_criteria_goal=filter_criteria_start,\n",
    "    device=device,\n",
    "    cells_to_generate=None,\n",
    "    top_genes=gene_viz_set,\n",
    "    latent=False\n",
    ")\n",
    "propagated_cells_2.columns = list(adata.var.index.copy()) + [\"perturbation\"]\n",
    "\n",
    "\n",
    "# Plot both conditions side by side\n",
    "densities=plot_density_grid(\n",
    "    propagated_cells_2, \n",
    "    propagated_cells_1, \n",
    "    gene_viz_set, \n",
    "    \"perturbation\",\n",
    "    title2=\"Beta cell: Control to T2D\",\n",
    "    title1=\"Beta cell: T2D to Control\"\n",
    ")\n",
    "\n",
    "densities.savefig('hpap/densities_curated.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d1abcd8-7c87-41fb-b28e-5b03d410dc5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ccf68e4-e0e5-4797-afc8-c35fad4d7dba",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "propagated_cells=generate_examples(adata=adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                          propagator=propagator,\n",
    "                                          shared_filter=shared_filter,\n",
    "                                 filter_criteria_start=filter_criteria_start,\n",
    "                                 filter_criteria_goal=filter_criteria_goal,\n",
    "                                 device=device,cells_to_generate=2000,\n",
    "                                   top_genes=list(perturbation_metrics.sort_values(\"oracle_score\",ascending=False).tail(3)['label']),\n",
    "                                   latent=True\n",
    "                                  )\n",
    "\n",
    "print(propagated_cells.shape)\n",
    "umap_data=pca_umap_transform(dataframe=propagated_cells, key_column='key', n_pca_components=25, n_umap_components=2, n_neighbors=5, min_dist=1)\n",
    "\n",
    "plot_umap(\n",
    "    dataframe=umap_data,\n",
    "    key_column=\"key\",\n",
    "    highlight_keys=list(perturbation_metrics.sort_values(\"oracle_score\",ascending=False).tail(3)['label'])+['start_prop','goal_prop'],  # Highlight key1 and key2\n",
    "    figsize=(8, 6),\n",
    "    default_color=\"lightgray\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d8bd6fa-4430-41f3-8696-c5ac7766467e",
   "metadata": {},
   "outputs": [],
   "source": [
    "propagated_cells=generate_examples(adata=adata,\n",
    "                                 stargan=stargan,\n",
    "                                 oracle=oracle,\n",
    "                                          propagator=propagator,\n",
    "                                          shared_filter=shared_filter,\n",
    "                                 filter_criteria_start=filter_criteria_start,\n",
    "                                 filter_criteria_goal=filter_criteria_goal,\n",
    "                                 device=device,cells_to_generate=2000,\n",
    "                                   top_genes=list(perturbation_metrics.sort_values(\"oracle_score\",ascending=False).head(3)['label']),\n",
    "                                   latent=False\n",
    "                                  )\n",
    "\n",
    "print(propagated_cells.shape)\n",
    "umap_data=pca_umap_transform(dataframe=propagated_cells, key_column='key', n_pca_components=25, n_umap_components=2, n_neighbors=10, min_dist=1)\n",
    "\n",
    "plot_umap(\n",
    "    dataframe=umap_data,\n",
    "    key_column=\"key\",\n",
    "    highlight_keys=list(perturbation_metrics.sort_values(\"oracle_score\",ascending=False).head(2)['label'])+['start_prop','goal_prop'],  # Highlight key1 and key2\n",
    "    figsize=(8, 6),\n",
    "    default_color=\"lightgray\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f95b4cd7-a9e0-4b8f-a9ab-21476400ee59",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "# Horizontal bar plot\n",
    "pm=perturbation_metrics.sort_values('oracle_score',ascending=False)#.tail(100)\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create the scatter plot with transparency\n",
    "plt.figure(figsize=(20, 8))\n",
    "plt.scatter(x=pm['label'], y=pm['oracle_score'], alpha=0.3, color='blue', s=10)\n",
    "\n",
    "# Add title and labels\n",
    "plt.title(\"Scores of 2000 Individuals\", fontsize=16)\n",
    "plt.xlabel(\"Label\", fontsize=14)\n",
    "plt.ylabel(\"Oracle Score\", fontsize=14)\n",
    "plt.xticks(rotation=90)\n",
    "plt.axhline(y=0.0,linewidth=1, color='red')\n",
    "# Improve layout\n",
    "plt.tight_layout()\n",
    "plt.grid(False)\n",
    "# Show the plot\n",
    "plt.show()\n",
    "\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
